teps/0093-add-sign-verify-subcommand-to-the-cli.md

```diff
@@ -0,0 +1,161 @@
+
+**Trusted Resources Users**: [TEP-0091](https://github.com/tektoncd/community/blob/main/teps/0091-trusted-resources.md) introduces the verification at pipeline controller but doesn't cover the signing. This tep can provide a tool for Trusted Resources Users to sign their resources.
+
+### Requirements
+
+* Signing a resource shouldn't change the content of it, e.g. we should not call `Setdefaults` for these resources to add default values.
+
+## Proposal
+
+The `sign` and `verify` subcommands should be added to current `task` and `pipeline` commands.  Each subcommand has flags. Details will be discussed in next section.
+
+## Design Details
+
+### Command line specifications
+**Sign:**
+```shell
+❯ tkn task sign --help
+Sign Tekton Task
+
+Usage:
+tkn task sign [flags]
+
+Examples:
+ Sign a Task from yaml file and save the signed Task:
+ tkn task sign examples/example-task.yaml  -K=cosign.key -f=signed.yaml
+
+Flags:
+  -h, --help        help for sign
+  -K, --keyFile     private key file path
+  -m, --kmsKey      kms key path
+  -f, --targetFile  Filename of the signed file
+```
+
+**Verify:**
+```shell
+❯ tkn task verify --help
+Verify Tekton Task
+
+Usage:
+tkn task verify [flags]
+
+Examples:
+ Verify a signed Task file:
+ tkn task verify examples/signed.yaml  -K=cosign.pub
+
+Flags:
+  -h, --help        help for verify
+  -K, --keyFile     public key file path
+  -m, --kmsKey      kms key path
+```
+
+### Sign
+To sign a resource file:
+```bash
+tkn task sign examples/example-task.yaml  -K=cosign.key -f=signed.yaml
+```
+Flags:
+
+|  Flag | ShortFlag  |  Description  |
+|---|---|---|
+| keyFile  | K   |  private key file path |
+| kmsKey  | m   | kms key path  |
+| targetFile  |  f | Filename of the signed file  |
+
+This will read private and unmarshal yaml files to get Tekton CRD (Task/Pipeline), signing function should sign the hashed bytes of the CRD, attach the base64 encoded signature to annotation with key as `tekton.dev/signature`.
+
+### Verify
+To verify a file:
+```bash
+tkn task verify signed.yaml -K=cosign.pub -d=Task
+```
+
+|  Flag | ShortFlag  |  Description  |
+|---|---|---|
+| keyFile  | K   |  public key file path |
+| kmsKey  | m   | kms key path  |
+
+Logs in terminal should tell users whether verification succeed or not.
+
+## Alternatives
+
+1. Cosign cli
+Cosign supports signing [tekton bundles](https://github.com/sigstore/cosign#tekton-bundles) and other type of data. But Cosign cli is out of Tekton ecosystem it would be better to have feature implemented into Tekton CLI.
+
+2. tkn plugins
+These functions can be stored in a separate repo and installed as tkn plugins. But that is not good for user experience.
```

Comment:
I guess it depends on the plugin distribution infrastructure, but at the same time it's best to keep it in tree unless there is a special need for a plugin.

Reply:
I think plugin is an alternative in this proposal?